akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -363,17 +368,13 @@ import org.reactivestreams.Subscriber
 
       private val callback = getAsyncCallback[Output[T]] {
         case QueueSink.Pull(pullPromise) =>
-          currentRequest match {
-            case Some(_) =>
-              pullPromise.failure(
-                new IllegalStateException(
-                  "You have to wait for previous future to be resolved to send another request"))
-            case None =>
-              if (buffer.isEmpty) currentRequest = Some(pullPromise)
-              else {
-                if (buffer.used == maxBuffer) tryPull(in)
-                sendDownstream(pullPromise)
-              }
+          if (currentRequests.size >= requestBufferSize)
+            pullPromise.failure(new IllegalStateException(
+              "Request buffer is full. You have to wait for one previous future to be resolved to send another request"))
```

Comment:
That there is a request buffer is an implementation concern, what is really happening is that there are too many concurrent requests. I think the error message could also include the buffer size as the max number of concurrent `SinkQueue.pull`s.

Reply:
Fixed